### PR TITLE
ansible-galaxy - enable remote requirements file processing

### PR DESCRIPTION
**Pull request by eryx12o45**
##### SUMMARY
Currently it is not possible to install roles and collections from a remote requirements.yml file. Only local files are allowed. This feature extends the option to specify an url to an requirements.yml file and parses the remote source. This allows users a central requirements management for multiple ansible projects.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
ansible galaxy cli

### CODE REVIEW

File: lib/ansible/cli/galaxy.py [hunk 1]
```diff
@@ -12,7 +12,6 @@
 import sys
 import textwrap
 import time
-
```
Comment (webknjaz) [inline lib/ansible/cli/galaxy.py:15]: This empty line was a separator between stdlib imports and the third-party ones. Please revert any formatting changes as they are unrelated to what you claim to patch in the PR title.

File: test/integration/targets/ansible-galaxy-role/tasks/main.yml [hunk 2]
```diff
@@ -56,3 +56,14 @@
 
 - name: Uninstall invalid role
   command: ansible-galaxy role remove invalid-testrole
+
+- name: Install invalid role from remote requirements.yml
+  command: ansible-galaxy install -r https://github.com/eryx12o45/ansible/tree/add_new_galaxy_option/test/integration/targets/ansible-galaxy-role/files/invalid-requirements.yml
```
Comment (webknjaz) [inline test/integration/targets/ansible-galaxy-role/tasks/main.yml:61]: This would need to be a URL to some of the repositories controlled by us. We cannot rely on arbitrary content on the internet.

File: lib/ansible/cli/galaxy.py [hunk 3]
```diff
@@ -52,6 +51,13 @@
 from ansible.utils.display import Display
 from ansible.utils.plugin_docs import get_versioned_doclink
 
+try:
+    # python3
+    import urllib.request as urllib
+except ImportError:
+    # python2
+    import urllib2 as urllib
```
Comment (webknjaz) [inline lib/ansible/cli/galaxy.py:59]: As mentioned by the linter comments, the vendored lib must be used.

**Review by webknjaz — CHANGES_REQUESTED** [review]
I'm personally, :-1: on this idea because:
- it complicates otherwise simple execution flow
- it opens up a can of worms security-wise: the URLs are not always idempotent and their content may be hijacked
  * if somebody *wants* to use this sort of setup, they could easily `wget -O requirements.yml https://some.url.com/file && ansible-galaxy install -r requirements.yml`; although, I wouldn't recommend doing so because it hurts reproducibility
  * instead, I encourage you to use Git subtrees (or maybe Git submodules, though I'd stick with subtrees personally) and merge-in this file from the central repository into the role ones periodically (`git merge --allow-unrelated-histories` is your friend here) to ensure that you actually have at least some idempotency in your system
- this is not something ansible-core should deal with and I'd rather not encourage end-users to make their automations more fragile by adding such a feature
- I haven't seen any other package manager having a similar feature, have you? I'm pretty sure there's a good reason behind that.

But hey, let's see what others think before closing this for good.

File: test/integration/targets/ansible-galaxy-role/tasks/main.yml (hunk 2 above)
Comment (eryx12o45) [reply test/integration/targets/ansible-galaxy-role/tasks/main.yml:61]: Sure thing, just for the imitial testing it had to be a valid url. I can change this nevertheless once there is a valid file

**Review by samdoran — CHANGES_REQUESTED** [review]
I am also 👎 to this change mostly due to the potential for hijacking the remote requirements files. This is something that can be done outside of Ansible, but is not a behavior that should be encouraged by building it into Ansible itself.

**Review by s-hertel — CHANGES_REQUESTED** [review]
I agree with  the other comments and am -1 on including this feature. Rather than overload ansible-galaxy, you can just use a preceding task that downloads the requirements file.
